web/src/engine/osk/src/input/gestures/specsForKeyboard.ts

```diff
@@ -0,0 +1,349 @@
+  itemPriority: 0,
+  itemChangeAction: 'reject',
+  pathResolutionAction: 'resolve',
+  pathModel: {
+    evaluate: (path) => {
+      if(path.isComplete && !path.wasCancelled) {
+        return 'resolve';
+      }
+    }
+  }
+}
+
+export const SubkeySelectContactModel: ContactModel = {
+  itemPriority: 0,
+  pathResolutionAction: 'resolve',
+  pathModel: {
+    evaluate: (path) => {
+      if(path.isComplete && !path.wasCancelled) {
+        return 'resolve';
+      }
+    }
+  }
+}
+// #endregion
+
+// #region Gesture-stage model definitions
+type GestureModel = specs.GestureModel<KeyElement>;
+
+// TODO:  customization of the gesture models depending upon properties of the keyboard.
+// - has flicks?  no longpress shortcut, also no longpress reset(?)
+// - modipress:  keyboard-specific modifier keys - which may require inspection of a
+//   key's properties.
+
+// Is kind of a mix of the two longpress styles.
+export const LongpressModel: GestureModel = {
+  id: 'longpress',
+  resolutionPriority: 0,
+  contacts: [
+    {
+      model: {
+        // Is the version without the up-flick shortcut.
+        ...MainContactLongpressSourceModel,
+        itemPriority: 1,
+        pathInheritance: 'chop'
+      },
+      endOnResolve: false
+    }, {
+      model: InstantContactRejectionModel
+    }
+  ],
+  resolutionAction: {
+    type: 'chain',
+    next: 'subkey-select',
+    selectionMode: 'none',
+    item: 'none'
+  },
+  /*
+   * Note:  these actions make sense in a 'roaming-touch' context, but not when
+   * flicks are also enabled.
+   */
+  rejectionActions: {
+    item: {
+      type: 'replace',
+      replace: 'longpress'
+    },
+    path: {
+      type: 'replace',
+      replace: 'longpress'
+    }
+  }
+}
+
+export const MultitapModel: GestureModel = {
+  id: 'multitap',
+  resolutionPriority: 2,
+  contacts: [
+    {
+      model: {
+        ...SimpleTapContactModel,
+        itemPriority: 1,
+        pathInheritance: 'reject',
+        allowsInitialState(incomingSample, comparisonSample, baseItem) {
+          // By default, the state token is set to whatever the current layer is for a source.
+          //
+          // So, if the first tap of a key swaps layers, the second tap will be on the wrong layer and
+          // thus have a different state token.  This is the perfect place to detect and correct that.
+          if(comparisonSample.stateToken != incomingSample.stateToken) {
+            incomingSample.stateToken = comparisonSample.stateToken;
+
+            // TODO: specialized item lookup required here for proper 'correction', corresponding to
+            // the owning VisualKeyboard.  That rigging doesn't exist quite yet, at the time of writing this.
+            incomingSample.item = undefined;
+          }
+          return incomingSample.item == baseItem;
+        },
+      },
+      endOnResolve: true
+    }, {
+      model: InstantContactResolutionModel
+    }
+  ],
+  sustainTimer: {
+    duration: 500,
```

Comment:
I think we need to surface the magic timing constants somewhere together

Reply:
Agreed.  The start of that is currently in #9698.  Multitap stuff is a follow-up to that - #9740 - so I'll aim to make sure it's handled there.